File: WordSquare/src/WordSquare.java
import java.lang.reflect.Array;
import java.util.ArrayList;

public class WordSquare {
    public ArrayList<String> words;
    public ArrayList<String> matchedWords = new ArrayList<String>();

    public WordSquare(ArrayList<String> words) {
        this.words = words;
    }

    public boolean process() {
        this.matchedWords.add(this.words.get(0));
        this.words.remove(0);

        for (String word: words) { // ball, area, lead, food, lady
            if (this.matchedWords.size() < 4) {
                checkWords(matchedWords, words, matchedWords.size()); // [ball], [area, lead, lady], 1
            }
        }

        if (matchedWords.size() == 4) {
            System.out.println("A Wordsquare can be created with: ");
            for (String match: matchedWords) {
                System.out.println(match);
            }
            return true;
        } else {
            return false;
        }
    }

    public boolean checkWords(ArrayList<String> matchedWords, ArrayList<String> inputWords, Integer position) {
        String searchString = searchString(this.matchedWords, position); // a

        for (String inputWord: inputWords) {
            if (this.matchedWords.contains(inputWord)) {
                continue;
            }

            String inputWordSub = inputWord.substring(0, this.matchedWords.size());

            if (inputWordSub.equals(searchString)) {
                this.matchedWords.add(inputWord);
                return true;
            } else {
                continue;
            }
        }

        return false;
    }

    public String searchString(ArrayList<String> matchedWords, Integer position) {
        String search = "";

        for (String matchedWord: matchedWords) {
            String subS = matchedWord.substring(position, position + 1);
            search += subS;
        }

        return search;
    }
}

File: scala/WordSquare/1/src/main/scala/WordSquare.scala
package com.mikekrisher.word_square

import util.control.Breaks._

class WordSquare(words: List[String]) {
  var matchedWords : List[String] = List()

  def process : Boolean = 
  {
    matchedWords = List(matchedWords ++ List(words(0))).flatten

    val remainingWords = words.drop(1)

    for (i <- 0 to (remainingWords.size - 1)) 
    {
      if (matchedWords.size < 4) {
        val wordMatch = checkWords(matchedWords, remainingWords, matchedWords.size)
      }

      if (matchedWords.size == 4) {
        println("A WordSquare can be created with: ")
        for (j <- 0 to (matchedWords.size - 1))
        {
          println(matchedWords(j))
        }
        return true
      } 
    }

    return false
  }

  def checkWords(matchingWords: List[String], inputWords: List[String], position: Int) : Boolean = 
  {
    val search = searchString(matchingWords, position)

    for (i <- 0 to (inputWords.size - 1))
    {
      val inputWord = inputWords(i)

      breakable {
        if (matchingWords.contains(inputWord)) {
          break
        }
      }

      val inputWordSub = inputWord.substring(0, matchingWords.size)

      breakable {
        if (inputWordSub == search) {
          matchedWords = List(matchedWords ++ List(inputWord)).flatten
          return true
        }
      }
    }

    return false
  }

  def searchString(matchingWords: List[String], position: Int) : String =
  {
    var search = "";

    for (i <- 0 to (matchingWords.size - 1)) 
    {
      val subString = matchingWords(i).substring(position, position + 1)
      search = search.concat(subString)
    }

    return search
  }
}

